lib/items/facebook.dart

```diff
@@ -5,55 +5,64 @@ import 'package:html/parser.dart' show parse;
 import 'item.dart';
 
 class FacebookItem extends MultaccItem {
-  String username;
   String userId;
 
+  String _username;
+  get username => _username ?? '';
+  set username(String input) {
+    _username = input;
+    _fetchId();
+  }
+
   FacebookItem();
 
   FacebookItem.fromJson(Map<String, dynamic> json)
-      : username = json['username'],
+      : _username = json['username'],
         userId = json['id'];
 
-  toMap() => {'username': username, 'userId': userId};
+  toMap() => {'username': username, 'id': userId};
+
+  get humanReadableValue => username != '' ? username : (userId != null ? 'fb.com/$userId' : '');
 
-  get humanReadableValue => username ?? (userId != null ? 'fb.com/$userId' : '');
+  get _validUsernameOrId => username != '' ? username : userId;
 
   get type => MultaccItemType.Facebook;
 
   launchApp() async {
     try {
       await launch('fb://profile/$userId');
     } catch (ex) {
-      launch('https://facebook.com/${username ?? userId}');
+      launch('https://facebook.com/$_validUsernameOrId');
     }
   }
 
   launchMessenger() async {
     try {
-      await launch('fb-messenger://user/${username ?? userId}');
+      await launch('fb-messenger://user/$_validUsernameOrId');
     } catch (ex) {
-      launch('https://facebook.com/messages/t/${username ?? userId}');
+      launch('https://facebook.com/messages/t/$_validUsernameOrId');
     }
   }
 
   get isLaunchable => true;
 
+  // @todo Add 'value' getter for editing
   set value(String input) {
-    String domain = input.toLowerCase().split('/')[0];
-
-    if (domain.contains('fb.me') || domain.contains('m.me')) {
-      username = input.split('/').last;
-    } else if (input.contains('profile.php?id=')) {
-      userId = input.split('id=')[1];
-    } else if (domain.contains('facebook.com') || domain.contains('fb.com')) {
-      username = input.split('/').last;
+    Uri url = Uri.parse(input);
+    String urlText = url.toString().toLowerCase();
+
+    if (urlText.contains('fb.me') || urlText.contains('m.me')) {
+      username = url.pathSegments.last;
+    } else if (urlText.contains('profile.php')) {
+      userId = url.queryParameters['id'];
+      _username = '';
+    } else if (urlText.contains('facebook.com') || urlText.contains('fb.com')) {
+      username = url.pathSegments.last;
```

Comment:
As is this handles the following inputs wrong so cannot be merged:
https://facebook.com/100007027231184 (valid link)
https://fb.me/100007027231184 (invalid link for some fucking reason)
https://facebook.com/blam.merk/photos (username contains m.me; additional path segments after username)
https://facebook.com/iamaprofile.phphp (this is a valid username)

```suggestion
    Uri url;
    try {
      url = Uri.parse(input);
    } on FormatException {
      url = null;
    }
    String domain = url?.host ?? '';
    
    if (domain.startsWith('www.')) {
      domain = domain.substring(4);
    }
    if (url?.queryParameters?.containsKey('id') ?? false) {
      userId = url.queryParameters['id'];
      _fetchUsername();
    } else {
      String user = [
        'm.me', 
        'fb.me', 
        'fb.com', 
        'facebook.com',
      ].contains(domain) ? url.pathSegments[0] : input.trim();
      if (domain != 'fb.me' && RegExp(r'\d+').hasMatch(user)) {
        userId = user;
        _fetchUsername();
      } else {
        username = user;
        _fetchId();
      }
    }
```